Cirrious/Cirrious.MvvmCross/ViewModels/MvxAsyncCommand.cs

```diff
@@ -0,0 +1,283 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Cirrious.CrossCore;
+using Cirrious.CrossCore.ExtensionMethods;
+using Cirrious.CrossCore.Platform;
+
+namespace Cirrious.MvvmCross.ViewModels
+{
+    public abstract class MvxAsyncCommandBase
+        : MvxCommandBase
+    {
+        private readonly object _syncRoot = new object();
+        private readonly bool _allowConcurrentExecutions;
+        private CancellationTokenSource _cts;
+        private int _concurrentExecutions;
+
+        protected MvxAsyncCommandBase(bool allowConcurrentExecutions = false)
+        {
+            _allowConcurrentExecutions = allowConcurrentExecutions;
+        }
+
+        public event EventHandler<MvxCommandErrorEventArgs> ErrorOccured;
+
+        public bool IsRunning
+        {
+            get { return _concurrentExecutions > 0; }
+        }
+
+        protected CancellationToken CancelToken 
+        {
+            get 
+            {
+                return _cts.Token; 
+            }
+        }
+
+        protected abstract bool DoCanExecute(object parameter);
+        protected abstract Task DoExecuteAsync(object parameter);
+
+        public void Cancel()
+        {
+            lock (_syncRoot)
+            {
+                if (_cts == null)
+                {
+                    Mvx.Trace(MvxTraceLevel.Warning, "MvxAsyncCommand : Attempt to cancel a task that is not running");
+                }
+                else
+                {
+                    _cts.Cancel();
+                }
+            }
+        }
+
+        public bool CanExecute()
+        {
+            return CanExecute(null);
+        }
+
+        public bool CanExecute(object parameter)
+        {
+            if (!_allowConcurrentExecutions && IsRunning)
+                return false;
+            else
+                return DoCanExecute(parameter);
+        }
+
+        public async void Execute(object parameter)
+        {
+            await ExecuteAsync(parameter).ConfigureAwait(false);
+        }
+
+        public async void Execute()
```

Comment:
I think we need the synchronous (fire and forget) version so it can be bound.
That being said, you are right : if we want custom exception handling on async command we should also catch exceptions from `CanExecute` and chose what to do with it.
But that's another argument against the custom exception handling I implemented. Moreover, nothing special was done for synchronous command. If an exception is raised by your CanExecute (or Execute) while it's used from binding, I think it'll crash the application and that's expected.